### PR TITLE
Remove metrics CRD

### PR DESCRIPTION
**Pull request by kalantar**
Addresses issues https://github.com/iter8-tools/iter8/issues/930, https://github.com/iter8-tools/iter8/issues/934, and https://github.com/iter8-tools/iter8/issues/973.

### CODE REVIEW

File: api/v2beta1/experiment_types.go [hunk 1]
```diff
@@ -73,16 +73,12 @@ type ExperimentSpec struct {
...
-	// Name is identifier for metric.  Can be of the form "name" or "namespace/name"
-	Name string `json:"name" yaml:"name"`
+	// Backend is details of metrics backend or overrides to a backend defined in a configmap
+	Backends []Backend `json:"backends,omitempty" yaml:"backends,omitempty"`
 
-	// MetricObj is the referenced metric
-	// +kubebuilder:validation:EmbeddedResource
-	MetricObj Metric `json:"metricObj" yaml:"metricObj"`
+	// Metrics are metrics definitions used by Criteria; may override predefined metrics
+	Metrics []Metric `json:"metrics,omitempty" yaml:"metrics,omitempty"`
```
Comment (sriumcp) [inline api/v2beta1/experiment_types.go:81]: This should go.

**Review by sriumcp — CHANGES_REQUESTED** [review]
Change requested for metrics (which appear twice)

File: api/v2beta1/experiment_types.go (hunk 1 above)
Comment (kalantar) [reply api/v2beta1/experiment_types.go]: done

File: api/v2beta1/defaults.go [hunk 2]
```diff
@@ -31,9 +31,6 @@ const (
 	// DefaultFailureHandler is the prefix of the default failure handler
 	DefaultFailureHandler string = "finish"
```
Comment (sriumcp) [inline api/v2beta1/defaults.go:32]: We really don't have a concept of a failure handler. It is `start`, `loop` and `finish`. 

Could we please  remove FailureHandler related code please?

**Review by sriumcp — COMMENTED** [review]
We also said that we will *not* retain unknown fields. And CR cannot contain unknown fields. Is this how kubebuilder annotations are set now?

File: test/data/withbackend.yaml [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+    secret: default/my-secret
+    headers:
+    - name: header
+      value: "{{.variable-1}}::{{.variable-2}}"
+    url: https://provider.url
+    metrics:
+    - name: "backend-1/reward-metric"
+      description: "reward-metric description"
+      params:
+      - name: params
```
Comment (sriumcp) [inline test/data/withbackend.yaml:29]: This looks super clean.

**Review by sriumcp — COMMENTED** [review]
We can also get rid of `NamedValue`.

I see that they are still used in two places.

1. Params for metrics, and 
2. Headers for metrics.

We can make them both maps.

This is related to your query the other day: are there really multiple params? Definitely not often. Probably infrequently. Anyway, we can cut out a line from both headers and params if we use maps instead of name-value pairs.

File: api/v2beta1/defaults.go (hunk 2 above)
Comment (kalantar) [reply api/v2beta1/defaults.go]: done

**Comment by kalantar** [comment]
> We also said that we will not retain unknown fields. And CR cannot contain unknown fields. Is this how kubebuilder annotations are set now?

The default behavior is now implemented. We don't have any directives related to this.  The only place we find the property `x-kubernetes-preserve-unknown-fields: true` is in the generated crd file as a result of using:

 ```yaml
With map[string]apiextensionsv1.JSON
```

**Review by sriumcp — CHANGES_REQUESTED** [review]
1. Could we please remove comments related to target?  For example, stage initializing is now purely about start handler, not target acquisition: https://github.com/kalantar/etc3/blob/af921b1494b99d60e2f89fdae86a5775b9b7e60e/api/v2beta1/constants.go#L88

2. Could we please remove `MetricUnavailable` and `MetricUnreadable` reason codes from constants? Metric related validation is extremely important, and we need to provide proper support for it. But we need to think through how to achieve this comprehensively **within the analytics service**.  https://github.com/kalantar/etc3/blob/af921b1494b99d60e2f89fdae86a5775b9b7e60e/api/v2beta1/constants.go#L67

**Comment by kalantar** [comment]
You managed to find the only use of the word "target" !  Fixed.

**Review by sriumcp — CHANGES_REQUESTED** [review]
This PR is almost there. I have two comments... 

1. We recently had an issue with `v2alpha` `run` task. Essentially, if both *task* and *run* are unspecified, the controller fails at the target acquisition stage, with no indication of why it was failing. Could we please introduce this check during validation (i.e., each task must contain exactly one of task or run but not both)? If it fails the check, we can fail the experiment with a *really understandable* failure code / reason. This would help in authoring as well as debugging.

2. While going through the handler parts of the controller logic, I saw a *commented out* portion that hardcodes the handler job structure (i.e., defines a batch/v1 job struct) instead of reading from the configmap. I believe this  is an extremely useful  alternative solution for what we currently have. This would basically knockout the whole handler configmap during installation, which is very desirable. Could we enable this approach? The only variable here would be handler image, which we can set as an environment variable for etc3.

**Comment by kalantar** [comment]
Created issues https://github.com/iter8-tools/iter8/issues/982 and https://github.com/iter8-tools/iter8/issues/983 to address the suggested additional items. I would rather not add to this PR.

**Review by sriumcp — APPROVED** [review]
lgtm

**Pull request merged**